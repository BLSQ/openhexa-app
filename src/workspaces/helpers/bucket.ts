import { gql } from "@apollo/client";
import { getApolloClient } from "core/helpers/apollo";
import {
  GetFileDownloadUrlMutation,
  GetFileDownloadUrlMutationVariables,
  DeleteBucketObjectMutation,
  DeleteBucketObjectMutationVariables,
  GetBucketUploadUrlMutation,
  GetBucketUploadUrlMutationVariables,
} from "./bucket.generated";

export async function getBucketObjectDownloadUrl(
  workspaceSlug: string,
  objectKey: string,
): Promise<string> {
  const client = getApolloClient();
  const { data } = await client.mutate<
    GetFileDownloadUrlMutation,
    GetFileDownloadUrlMutationVariables
  >({
    mutation: gql`
      mutation GetFileDownloadUrl($input: PrepareObjectDownloadInput!) {
        prepareObjectDownload(input: $input) {
          success
          downloadUrl
        }
      }
    `,
    variables: {
      input: {
        workspaceSlug,
        objectKey,
      },
    },
  });

  if (data?.prepareObjectDownload?.success) {
    return data.prepareObjectDownload.downloadUrl as string;
  } else {
    throw new Error("Object cannot be downloaded");
  }
}

export async function downloadURL(url: string, target: string = "") {
  const anchor = document.createElement("a");
  anchor.href = url;
  anchor.target = target;
  document.body.appendChild(anchor);
  anchor.click();
  document.body.removeChild(anchor);
}

export async function deleteBucketObject(
  workspaceSlug: string,
  objectKey: string,
) {
  const client = getApolloClient();
  const { data } = await client.mutate<
    DeleteBucketObjectMutation,
    DeleteBucketObjectMutationVariables
  >({
    mutation: gql`
      mutation deleteBucketObject($input: DeleteBucketObjectInput!) {
        deleteBucketObject(input: $input) {
          success
          errors
        }
      }
    `,
    variables: { input: { workspaceSlug, objectKey } },
    update(cache) {
      const normalizedId = cache.identify({
        key: objectKey,
        __typename: "BucketObject",
      });

      cache.evict({ id: normalizedId });
      cache.gc();
    },
  });

  if (!data?.deleteBucketObject.success) {
    throw new Error("Unexpected error.");
  }
}

export async function getBucketObjectUploadUrl(
  workspaceSlug: string,
  key: string,
  contentType: string,
) {
  const client = getApolloClient();
  const { data } = await client.mutate<
    GetBucketUploadUrlMutation,
    GetBucketUploadUrlMutationVariables
  >({
    mutation: gql`
      mutation GetBucketUploadUrl($input: PrepareObjectUploadInput!) {
        prepareObjectUpload(input: $input) {
          success
          uploadUrl
        }
      }
    `,
    variables: {
      input: {
        workspaceSlug,
        objectKey: key,
        contentType,
      },
    },
  });

  if (data?.prepareObjectUpload?.success) {
    return data.prepareObjectUpload.uploadUrl as string;
  } else {
    throw new Error("Object cannot be uploaded");
  }
}

export async function createBucketFolder(
  workspaceSlug: string,
  folderKey: string,
) {
  const client = getApolloClient();
  const { data } = await client.mutate({
    mutation: gql`
      mutation CreateBucketFolder($input: CreateBucketFolderInput!) {
        createBucketFolder(input: $input) {
          success
          errors
          folder {
            key
            name
            type
          }
        }
      }
    `,
    variables: {
      input: {
        workspaceSlug,
        folderKey,
      },
    },
  });

  if (!data?.createBucketFolder.success) {
    throw new Error("Unexpected error.");
  }

  return data.createBucketFolder.folder;
}
