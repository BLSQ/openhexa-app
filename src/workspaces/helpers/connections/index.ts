import { ConnectionType } from "graphql-types";
export { convertFieldsToInput } from "./utils";
export type { ConnectionForm, FieldForm } from "./utils";

import postgresql from "./postgresql";
import { ConnectionDefinition } from "./utils";
import custom from "./custom";
import s3 from "./s3";
import gcs from "./gcs";
import iaso from "./iaso";
import dhis2 from "./dhis2";

const Connections: { [key in ConnectionType]: ConnectionDefinition } = {
  [ConnectionType.Postgresql]: postgresql,
  [ConnectionType.Custom]: custom,
  [ConnectionType.S3]: s3,
  [ConnectionType.Gcs]: gcs,
  [ConnectionType.Iaso]: iaso,
  [ConnectionType.Dhis2]: dhis2,
};

export default Connections;
