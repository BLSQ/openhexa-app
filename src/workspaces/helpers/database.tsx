type lang = "R" | "PYTHON";

export const getReadTableSnippet = (tableName: string) => {
  return `import os
import pandas as pd
from sqlalchemy import create_engine

engine = create_engine(os.environ["WORKSPACE_DATABASE_URL"])

pd.read_sql("SELECT * FROM ${tableName}", con=engine)`;
};

export const getUsageSnippet = (tableName: string, lang: lang) => {
  let text = "";
  switch (lang) {
    case "PYTHON":
      text = `import os
import pandas as pd
from sqlalchemy import create_engine

engine = create_engine(os.environ["WORKSPACE_DATABASE_URL"])

# Create sample dataframe
df = pd.DataFrame({"name": ["Jane", "John", "Tyler"], "age": [19, 17, 22]})

# Write data
df.to_sql("${tableName}", con=engine, if_exists="replace")`;
      break;
    case "R":
      text = `library(DBI)
  
con <- dbConnect(
    RPostgres::Postgres(),
    dbname = Sys.getenv("WORKSPACE_DATABASE_DB_NAME"),
    host = Sys.getenv("WORKSPACE_DATABASE_HOST"),
    port = Sys.getenv("WORKSPACE_DATABASE_PORT"),
    user = Sys.getenv("WORKSPACE_DATABASE_USERNAME"),
    password = Sys.getenv("WORKSPACE_DATABASE_PASSWORD")
)

dbWriteTable(con, "${tableName}", Data_fin, overwrite=TRUE)`;
      break;
  }
  return text;
};
