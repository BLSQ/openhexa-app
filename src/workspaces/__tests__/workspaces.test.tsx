import { render, screen } from "@testing-library/react";
import { TestApp } from "core/helpers/testutils";
import WorkspacePage from "pages/workspaces/[workspaceId]";
import { WorkspacePageDocument } from "workspaces/graphql/queries.generated";

describe("Workspaces", () => {
  it("renders the workspace page", async () => {
    const id = "a303ff37-644b-4080-83d9-a42bd2712f63";
    const graphqlMocks = [
      {
        request: {
          query: WorkspacePageDocument,
          variables: {
            id,
          },
        },
        result: {
          data: {
            workspace: {
              id: id,
              name: "Rwanda Workspace",
              description: "This is a description",
              countries: [],
              permissions: {
                update: false,
                delete: false,
                manageMembers: false,
              },
              members: {
                totalItems: 0,
                items: [],
              },
            },
          },
        },
      },
    ];

    const { container } = render(
      <TestApp mocks={graphqlMocks}>
        <WorkspacePage page={1} perPage={1} workspaceId={id} />
      </TestApp>
    );
    const elm = await screen.findByText("Rwanda Workspace", { selector: "a" });
    expect(elm).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("renders null as workspace doesn't exist", async () => {
    const id = "a303ff37-644b-4080-83d9-a42bd2712f63";
    const graphqlMocks = [
      {
        request: {
          query: WorkspacePageDocument,
          variables: {
            id,
          },
        },
        result: {
          data: {
            workspace: null,
          },
        },
      },
    ];

    const { container } = render(
      <TestApp mocks={graphqlMocks}>
        <WorkspacePage page={1} perPage={1} workspaceId={id} />
      </TestApp>
    );

    expect(container.firstChild).toBeNull();
    expect(container).toMatchSnapshot();
  });

  it("hides the edit button when user have not update permission ", async () => {
    const id = "a303ff37-644b-4080-83d9-a42bd2712f63";
    const graphqlMocks = [
      {
        request: {
          query: WorkspacePageDocument,
          variables: {
            id,
          },
        },
        result: {
          data: {
            workspace: {
              id: id,
              name: "Rwanda Workspace",
              description: "This is a description",
              permissions: {
                update: false,
                delete: false,
                manageMembers: false,
              },
              countries: [],
              members: {
                totalItems: 0,
                items: [],
              },
            },
          },
        },
      },
    ];

    const { container } = render(
      <TestApp mocks={graphqlMocks}>
        <WorkspacePage page={1} perPage={1} workspaceId={id} />
      </TestApp>
    );
    const editButton = screen.queryByText("Edit");
    expect(editButton).toBeNull();
    expect(container).toMatchSnapshot();
  });
});
