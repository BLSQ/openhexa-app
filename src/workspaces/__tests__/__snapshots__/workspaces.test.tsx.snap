// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Workspaces hides the edit button when user have not update permission  1`] = `<div />`;

exports[`Workspaces renders null as workspace doesn't exist 1`] = `<div />`;

exports[`Workspaces renders the workspace page 1`] = `
<div>
  <div
    class="flex min-h-screen w-screen"
  >
    <div
      class="fixed h-screen bg-gray-800 transition-all duration-75 w-64 2xl:w-72"
    >
      <div
        class="relative z-20 flex h-full flex-col"
      >
        <div
          class="flex h-full flex-grow flex-col border-r border-gray-200 bg-gray-800"
        >
          <div
            class="w-full"
          >
            <button
              class="group flex h-16 w-full items-center justify-center bg-gray-800 px-2 text-left hover:bg-gray-600"
            >
              <div
                class="line-clamp-2 flex-1 text-sm tracking-tight text-gray-50"
                title="Rwanda Workspace"
              >
                Rwanda Workspace
                <div
                  class="text-xs tracking-tighter text-gray-500 group-hover:text-gray-400"
                />
              </div>
              <svg
                aria-hidden="true"
                class="ml-1 h-4 w-4 text-gray-500 group-hover:text-gray-100"
                fill="currentColor"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M12.53 16.28a.75.75 0 01-1.06 0l-7.5-7.5a.75.75 0 011.06-1.06L12 14.69l6.97-6.97a.75.75 0 111.06 1.06l-7.5 7.5z"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
          <div
            class="mt-5 flex flex-grow flex-col"
          >
            <nav
              class="flex-1 space-y-1 px-0 pb-4"
            >
              <a
                class="text-md group relative flex items-center gap-3 px-2 py-2 font-medium  text-gray-300 hover:bg-gray-700 hover:text-white"
                href="/workspaces/test"
                rel="noopener noreferrer"
              >
                <div
                  class="absolute inset-y-0 left-0 w-1 bg-pink-500 transition-opacity opacity-0"
                />
                <svg
                  aria-hidden="true"
                  class="ml-1 h-5 w-5"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M2.25 12l8.954-8.955c.44-.439 1.152-.439 1.591 0L21.75 12M4.5 9.75v10.125c0 .621.504 1.125 1.125 1.125H9.75v-4.875c0-.621.504-1.125 1.125-1.125h2.25c.621 0 1.125.504 1.125 1.125V21h4.125c.621 0 1.125-.504 1.125-1.125V9.75M8.25 21h8.25"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                Home
              </a>
              <a
                class="text-md group relative flex items-center gap-3 px-2 py-2 font-medium  text-gray-300 hover:bg-gray-700 hover:text-white"
                href="/workspaces/test/files"
                rel="noopener noreferrer"
              >
                <div
                  class="absolute inset-y-0 left-0 w-1 bg-pink-500 transition-opacity opacity-0"
                />
                <svg
                  aria-hidden="true"
                  class="ml-1 h-5 w-5"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M3.75 9.776c.112-.017.227-.026.344-.026h15.812c.117 0 .232.009.344.026m-16.5 0a2.25 2.25 0 00-1.883 2.542l.857 6a2.25 2.25 0 002.227 1.932H19.05a2.25 2.25 0 002.227-1.932l.857-6a2.25 2.25 0 00-1.883-2.542m-16.5 0V6A2.25 2.25 0 016 3.75h3.879a1.5 1.5 0 011.06.44l2.122 2.12a1.5 1.5 0 001.06.44H18A2.25 2.25 0 0120.25 9v.776"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                Files
              </a>
              <a
                class="text-md group relative flex items-center gap-3 px-2 py-2 font-medium  text-gray-300 hover:bg-gray-700 hover:text-white"
                href="/workspaces/test/databases"
                rel="noopener noreferrer"
              >
                <div
                  class="absolute inset-y-0 left-0 w-1 bg-pink-500 transition-opacity opacity-0"
                />
                <svg
                  aria-hidden="true"
                  class="ml-1 h-5 w-5"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M20.25 6.375c0 2.278-3.694 4.125-8.25 4.125S3.75 8.653 3.75 6.375m16.5 0c0-2.278-3.694-4.125-8.25-4.125S3.75 4.097 3.75 6.375m16.5 0v11.25c0 2.278-3.694 4.125-8.25 4.125s-8.25-1.847-8.25-4.125V6.375m16.5 0v3.75m-16.5-3.75v3.75m16.5 0v3.75C20.25 16.153 16.556 18 12 18s-8.25-1.847-8.25-4.125v-3.75m16.5 0c0 2.278-3.694 4.125-8.25 4.125s-8.25-1.847-8.25-4.125"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                Database
              </a>
              <a
                class="text-md group relative flex items-center gap-3 px-2 py-2 font-medium  text-gray-300 hover:bg-gray-700 hover:text-white"
                href="/workspaces/test/connections"
                rel="noopener noreferrer"
              >
                <div
                  class="absolute inset-y-0 left-0 w-1 bg-pink-500 transition-opacity opacity-0"
                />
                <svg
                  aria-hidden="true"
                  class="ml-1 h-5 w-5"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M4.098 19.902a3.75 3.75 0 005.304 0l6.401-6.402M6.75 21A3.75 3.75 0 013 17.25V4.125C3 3.504 3.504 3 4.125 3h5.25c.621 0 1.125.504 1.125 1.125v4.072M6.75 21a3.75 3.75 0 003.75-3.75V8.197M6.75 21h13.125c.621 0 1.125-.504 1.125-1.125v-5.25c0-.621-.504-1.125-1.125-1.125h-4.072M10.5 8.197l2.88-2.88c.438-.439 1.15-.439 1.59 0l3.712 3.713c.44.44.44 1.152 0 1.59l-2.879 2.88M6.75 17.25h.008v.008H6.75v-.008z"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                Connections
              </a>
              <a
                class="text-md group relative flex items-center gap-3 px-2 py-2 font-medium  text-gray-300 hover:bg-gray-700 hover:text-white"
                href="/workspaces/test/pipelines"
                rel="noopener noreferrer"
              >
                <div
                  class="absolute inset-y-0 left-0 w-1 bg-pink-500 transition-opacity opacity-0"
                />
                <svg
                  aria-hidden="true"
                  class="ml-1 h-5 w-5"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M16.023 9.348h4.992v-.001M2.985 19.644v-4.992m0 0h4.992m-4.993 0l3.181 3.183a8.25 8.25 0 0013.803-3.7M4.031 9.865a8.25 8.25 0 0113.803-3.7l3.181 3.182m0-4.991v4.99"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                Pipelines
              </a>
            </nav>
          </div>
          <div
            class="mb-5 flex flex-shrink-0 flex-col items-center px-4"
          >
            <a
              class="flex h-8 items-center"
              href="/"
              rel="noopener noreferrer"
            >
              <img
                alt="OpenHEXA logo"
                class="h-full"
                src="/images/logo_with_text_white.svg"
              />
            </a>
          </div>
        </div>
        <button
          class="group absolute inset-y-0 right-0 border-r-4 border-transparent after:absolute after:inset-y-0 after:-left-1.5 after:block after:w-5 after:content-[''] hover:border-r-gray-500"
        >
          <div
            class="relative h-full"
          >
            <div
              class="pointer-events-none absolute inset-y-0 right-0 flex items-center"
            >
              <div
                class="pointer-events-auto invisible rounded-l-md bg-gray-500 p-1 pr-0.5 align-middle text-white group-hover:visible"
              >
                <svg
                  aria-hidden="true"
                  class="h-5 w-5"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M15.75 19.5L8.25 12l7.5-7.5"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
              </div>
            </div>
          </div>
        </button>
      </div>
    </div>
    <div
      class="w-full transition-all duration-75 pl-64 2xl:pl-72"
    >
      <main
        class="flex flex-1 flex-col transition-all mb-12"
      >
        <div
          class="sticky top-0 z-10 flex h-16 flex-shrink-0 items-center justify-between border-b border-gray-200 bg-white py-3 shadow group relative px-4 md:px-6 xl:px-10 2xl:px-12"
        >
          <div
            class="flex-1 flex items-center justify-between"
          >
            <nav
              aria-label="Breadcrumbs"
            >
              <ol
                class="flex items-center"
                role="list"
              >
                <li
                  class=""
                >
                  <div
                    class="flex items-center"
                  >
                    <span
                      class="text-sm hover:text-gray-700 font-medium text-gray-500"
                    >
                      <a
                        class=""
                        href="/workspaces/test"
                        rel="noopener noreferrer"
                      >
                        Rwanda Workspace
                      </a>
                    </span>
                  </div>
                </li>
              </ol>
            </nav>
          </div>
        </div>
        <div
          class="py-6 xl:py-8"
        >
          <div
            class="mx-auto px-4 md:px-6 xl:px-10  2xl:px-12"
          >
            <article
              class="sm:rounded-lg overflow-hidden shadow border-b border-gray-200 bg-white"
            >
              <section
                class="px-4 py-5 sm:px-6"
              >
                <div>
                  <div
                    class="prose prose-headings:font-medium max-w-3xl"
                  >
                    <p>
                      This is a description
                    </p>
                  </div>
                </div>
              </section>
            </article>
          </div>
        </div>
      </main>
    </div>
    <div
      class="fixed bottom-6 right-6"
    >
      <div
        class="relative"
        data-headlessui-state=""
      >
        <button
          aria-expanded="false"
          class="flex items-center outline-none"
          data-headlessui-state=""
          id="headlessui-popover-button-6fbe024f-2316-4265-a6e8-d65a837e308a"
          type="button"
        >
          <div
            class="flex h-12 w-12 items-center justify-center rounded-full bg-white text-3xl shadow-xl ring-1 ring-gray-500 ring-opacity-5 transition-all hover:bg-gray-50 hover:text-4xl"
          >
            ?
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Workspaces shows the jupyterhub entry when user have the launchNotebookServer permission  1`] = `
<div>
  <div
    class="flex min-h-screen w-screen"
  >
    <div
      class="fixed h-screen bg-gray-800 transition-all duration-75 w-64 2xl:w-72"
    >
      <div
        class="relative z-20 flex h-full flex-col"
      >
        <div
          class="flex h-full flex-grow flex-col border-r border-gray-200 bg-gray-800"
        >
          <div
            class="w-full"
          >
            <button
              class="group flex h-16 w-full items-center justify-center bg-gray-800 px-2 text-left hover:bg-gray-600"
            >
              <div
                class="line-clamp-2 flex-1 text-sm tracking-tight text-gray-50"
                title="Rwanda Workspace"
              >
                Rwanda Workspace
                <div
                  class="text-xs tracking-tighter text-gray-500 group-hover:text-gray-400"
                />
              </div>
              <svg
                aria-hidden="true"
                class="ml-1 h-4 w-4 text-gray-500 group-hover:text-gray-100"
                fill="currentColor"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M12.53 16.28a.75.75 0 01-1.06 0l-7.5-7.5a.75.75 0 011.06-1.06L12 14.69l6.97-6.97a.75.75 0 111.06 1.06l-7.5 7.5z"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
          <div
            class="mt-5 flex flex-grow flex-col"
          >
            <nav
              class="flex-1 space-y-1 px-0 pb-4"
            >
              <a
                class="text-md group relative flex items-center gap-3 px-2 py-2 font-medium  text-gray-300 hover:bg-gray-700 hover:text-white"
                href="/workspaces/a303ff37-644b-4080-83d9-a42bd2712f63"
                rel="noopener noreferrer"
              >
                <div
                  class="absolute inset-y-0 left-0 w-1 bg-pink-500 transition-opacity opacity-0"
                />
                <svg
                  aria-hidden="true"
                  class="ml-1 h-5 w-5"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M2.25 12l8.954-8.955c.44-.439 1.152-.439 1.591 0L21.75 12M4.5 9.75v10.125c0 .621.504 1.125 1.125 1.125H9.75v-4.875c0-.621.504-1.125 1.125-1.125h2.25c.621 0 1.125.504 1.125 1.125V21h4.125c.621 0 1.125-.504 1.125-1.125V9.75M8.25 21h8.25"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                Home
              </a>
              <a
                class="text-md group relative flex items-center gap-3 px-2 py-2 font-medium  text-gray-300 hover:bg-gray-700 hover:text-white"
                href="/workspaces/a303ff37-644b-4080-83d9-a42bd2712f63/files"
                rel="noopener noreferrer"
              >
                <div
                  class="absolute inset-y-0 left-0 w-1 bg-pink-500 transition-opacity opacity-0"
                />
                <svg
                  aria-hidden="true"
                  class="ml-1 h-5 w-5"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M3.75 9.776c.112-.017.227-.026.344-.026h15.812c.117 0 .232.009.344.026m-16.5 0a2.25 2.25 0 00-1.883 2.542l.857 6a2.25 2.25 0 002.227 1.932H19.05a2.25 2.25 0 002.227-1.932l.857-6a2.25 2.25 0 00-1.883-2.542m-16.5 0V6A2.25 2.25 0 016 3.75h3.879a1.5 1.5 0 011.06.44l2.122 2.12a1.5 1.5 0 001.06.44H18A2.25 2.25 0 0120.25 9v.776"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                Files
              </a>
              <a
                class="text-md group relative flex items-center gap-3 px-2 py-2 font-medium  text-gray-300 hover:bg-gray-700 hover:text-white"
                href="/workspaces/a303ff37-644b-4080-83d9-a42bd2712f63/databases"
                rel="noopener noreferrer"
              >
                <div
                  class="absolute inset-y-0 left-0 w-1 bg-pink-500 transition-opacity opacity-0"
                />
                <svg
                  aria-hidden="true"
                  class="ml-1 h-5 w-5"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M20.25 6.375c0 2.278-3.694 4.125-8.25 4.125S3.75 8.653 3.75 6.375m16.5 0c0-2.278-3.694-4.125-8.25-4.125S3.75 4.097 3.75 6.375m16.5 0v11.25c0 2.278-3.694 4.125-8.25 4.125s-8.25-1.847-8.25-4.125V6.375m16.5 0v3.75m-16.5-3.75v3.75m16.5 0v3.75C20.25 16.153 16.556 18 12 18s-8.25-1.847-8.25-4.125v-3.75m16.5 0c0 2.278-3.694 4.125-8.25 4.125s-8.25-1.847-8.25-4.125"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                Database
              </a>
              <a
                class="text-md group relative flex items-center gap-3 px-2 py-2 font-medium  text-gray-300 hover:bg-gray-700 hover:text-white"
                href="/workspaces/a303ff37-644b-4080-83d9-a42bd2712f63/connections"
                rel="noopener noreferrer"
              >
                <div
                  class="absolute inset-y-0 left-0 w-1 bg-pink-500 transition-opacity opacity-0"
                />
                <svg
                  aria-hidden="true"
                  class="ml-1 h-5 w-5"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M4.098 19.902a3.75 3.75 0 005.304 0l6.401-6.402M6.75 21A3.75 3.75 0 013 17.25V4.125C3 3.504 3.504 3 4.125 3h5.25c.621 0 1.125.504 1.125 1.125v4.072M6.75 21a3.75 3.75 0 003.75-3.75V8.197M6.75 21h13.125c.621 0 1.125-.504 1.125-1.125v-5.25c0-.621-.504-1.125-1.125-1.125h-4.072M10.5 8.197l2.88-2.88c.438-.439 1.15-.439 1.59 0l3.712 3.713c.44.44.44 1.152 0 1.59l-2.879 2.88M6.75 17.25h.008v.008H6.75v-.008z"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                Connections
              </a>
              <a
                class="text-md group relative flex items-center gap-3 px-2 py-2 font-medium  text-gray-300 hover:bg-gray-700 hover:text-white"
                href="/workspaces/a303ff37-644b-4080-83d9-a42bd2712f63/pipelines"
                rel="noopener noreferrer"
              >
                <div
                  class="absolute inset-y-0 left-0 w-1 bg-pink-500 transition-opacity opacity-0"
                />
                <svg
                  aria-hidden="true"
                  class="ml-1 h-5 w-5"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M16.023 9.348h4.992v-.001M2.985 19.644v-4.992m0 0h4.992m-4.993 0l3.181 3.183a8.25 8.25 0 0013.803-3.7M4.031 9.865a8.25 8.25 0 0113.803-3.7l3.181 3.182m0-4.991v4.99"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                Pipelines
              </a>
              <a
                class="text-md group relative flex items-center gap-3 px-2 py-2 font-medium  text-gray-300 hover:bg-gray-700 hover:text-white"
                href="/workspaces/a303ff37-644b-4080-83d9-a42bd2712f63/notebooks"
                rel="noopener noreferrer"
              >
                <div
                  class="absolute inset-y-0 left-0 w-1 bg-pink-500 transition-opacity opacity-0"
                />
                <svg
                  aria-hidden="true"
                  class="ml-1 h-5 w-5"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 6.042A8.967 8.967 0 006 3.75c-1.052 0-2.062.18-3 .512v14.25A8.987 8.987 0 016 18c2.305 0 4.408.867 6 2.292m0-14.25a8.966 8.966 0 016-2.292c1.052 0 2.062.18 3 .512v14.25A8.987 8.987 0 0018 18a8.967 8.967 0 00-6 2.292m0-14.25v14.25"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                JupyterHub
              </a>
            </nav>
          </div>
          <div
            class="mb-5 flex flex-shrink-0 flex-col items-center px-4"
          >
            <a
              class="flex h-8 items-center"
              href="/"
              rel="noopener noreferrer"
            >
              <img
                alt="OpenHEXA logo"
                class="h-full"
                src="/images/logo_with_text_white.svg"
              />
            </a>
          </div>
        </div>
        <button
          class="group absolute inset-y-0 right-0 border-r-4 border-transparent after:absolute after:inset-y-0 after:-left-1.5 after:block after:w-5 after:content-[''] hover:border-r-gray-500"
        >
          <div
            class="relative h-full"
          >
            <div
              class="pointer-events-none absolute inset-y-0 right-0 flex items-center"
            >
              <div
                class="pointer-events-auto invisible rounded-l-md bg-gray-500 p-1 pr-0.5 align-middle text-white group-hover:visible"
              >
                <svg
                  aria-hidden="true"
                  class="h-5 w-5"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M15.75 19.5L8.25 12l7.5-7.5"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
              </div>
            </div>
          </div>
        </button>
      </div>
    </div>
    <div
      class="w-full transition-all duration-75 pl-64 2xl:pl-72"
    >
      <main
        class="flex flex-1 flex-col transition-all mb-12"
      >
        <div
          class="sticky top-0 z-10 flex h-16 flex-shrink-0 items-center justify-between border-b border-gray-200 bg-white py-3 shadow group relative px-4 md:px-6 xl:px-10 2xl:px-12"
        >
          <div
            class="flex-1 flex items-center justify-between"
          >
            <nav
              aria-label="Breadcrumbs"
            >
              <ol
                class="flex items-center"
                role="list"
              >
                <li
                  class=""
                >
                  <div
                    class="flex items-center"
                  >
                    <span
                      class="text-sm hover:text-gray-700 font-medium text-gray-500"
                    >
                      <a
                        class=""
                        href="/workspaces/a303ff37-644b-4080-83d9-a42bd2712f63"
                        rel="noopener noreferrer"
                      >
                        Rwanda Workspace
                      </a>
                    </span>
                  </div>
                </li>
              </ol>
            </nav>
          </div>
        </div>
        <div
          class="py-6 xl:py-8"
        >
          <div
            class="mx-auto px-4 md:px-6 xl:px-10  2xl:px-12"
          >
            <article
              class="sm:rounded-lg overflow-hidden shadow border-b border-gray-200 bg-white"
            >
              <section
                class="px-4 py-5 sm:px-6"
              >
                <div>
                  <div
                    class="prose prose-headings:font-medium max-w-3xl"
                  >
                    <p>
                      This is a description
                    </p>
                  </div>
                </div>
              </section>
            </article>
          </div>
        </div>
      </main>
    </div>
    <div
      class="fixed bottom-6 right-6"
    >
      <div
        class="relative"
        data-headlessui-state=""
      >
        <button
          aria-expanded="false"
          class="flex items-center outline-none"
          data-headlessui-state=""
          id="headlessui-popover-button-6fbe024f-2316-4265-a6e8-d65a837e308a"
          type="button"
        >
          <div
            class="flex h-12 w-12 items-center justify-center rounded-full bg-white text-3xl shadow-xl ring-1 ring-gray-500 ring-opacity-5 transition-all hover:bg-gray-50 hover:text-4xl"
          >
            ?
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;
